src/mds/MDCache.cc

```diff
@@ -8515,8 +8515,11 @@ int MDCache::path_traverse(const MDRequestRef& mdr, MDSContextFactory& cf,
 	  lov.add_xlock(&dn->lock);
 	} else {
 	  // force client to flush async dir operation if necessary
-	  if (cur->filelock.is_cached())
+	  if (cur->filelock.is_cached() &&
+	      !(mdr->lock_cache &&
+		static_cast<const MutationImpl*>(mdr->lock_cache)->is_wrlocked(&cur->filelock))) {
 	    lov.add_wrlock(&cur->filelock);
+	  }
```

Comment:
Hmm, so it's correct that the current design of the lock caching is that we skip adding locks if held by the `lock_cache`.

What confuses me here is that this fix presupposes that the `mdr` is taking this `else` path. From what I can see in the log posted on the tracker, the `mdr` should be taking `if` path and not adding the `filelock` at all.

I do think this patch is fixing a genuine issue however. I think perhaps the right approach is:


```suggestion
	  if (cur->filelock.is_cached()) {
	    if (mdr->lock_cache)
	      mds->locker->put_lock_cache(mdr->lock_cache);
	    lov.add_wrlock(&cur->filelock);
	  }
```

Reply:
> Hmm, so it's correct that the current design of the lock caching is that we skip adding locks if held by the `lock_cache`.
> 
> What confuses me here is that this fix presupposes that the `mdr` is taking this `else` path. From what I can see in the log posted on the tracker, the `mdr` should be taking `if` path and not adding the `filelock` at all.
> 

This is because the `(dnl->is_null() || !want_inode)` was also `false` for `[inode 0x100244d7fcb [...2,head] /clusteradmin/weiler/deph-test/trash/`.

And then went to the `else` path.

> I do think this patch is fixing a genuine issue however. I think perhaps the right approach is:

I am afraid we will introduce other issues here. 

Because in the front of the `path_traverse()` it will attach the `lock_cache` to current `mdr` and then skip to call `try_rdlock_snap_layout()`:

```
 8362   if (flags & MDS_TRAVERSE_CHECK_LOCKCACHE)
 8363     mds->locker->find_and_attach_lock_cache(mdr, cur);                                                                                                                                                                                                  
 8364  
 8365   if (mdr && mdr->lock_cache) {
 8366     if (flags & MDS_TRAVERSE_WANT_DIRLAYOUT)
 8367       mdr->dir_layout = mdr->lock_cache->get_dir_layout();
 8368   } else if (rdlock_snap) {
 8369     int n = (flags & MDS_TRAVERSE_RDLOCK_SNAP2) ? 1 : 0; 
 8370     if ((n == 0 && !(mdr->locking_state & MutationImpl::SNAP_LOCKED)) ||
 8371         (n == 1 && !(mdr->locking_state & MutationImpl::SNAP2_LOCKED))) {
 8372       bool want_layout = (flags & MDS_TRAVERSE_WANT_DIRLAYOUT);
 8373       if (!mds->locker->try_rdlock_snap_layout(cur, mdr, n, want_layout))
 8374         return 1;
 8375     }
 8376   }
```

If we just drop the `lock_cache` here we should to do the above beforehand. Or retry the `path_traverse()`.

Reply:
And also I can be sure that the `lock_cache` feature will be disabled always, because we need to drop the `lock_cache` always.